DEP: linalg: sharpen deprecation warning for pinv {,r}cond

j-bowhay opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
<!--Please explain your changes.-->

Adds a specific version for removing `cond`/`rcond` and, whilst unlikely to be needed as the arguments are at the end of the signature, uses `_NoValue` to trigger the deprecation warning instead of `None`. Also adds a test which checks the deprecation warning is emitted.

#### Additional information
<!--Any additional information you think is important.-->

h-vetinari commented on scipy/linalg/_basic.py:
```diff
@@ -1435,14 +1436,15 @@ def pinv(a, atol=None, rtol=None, return_rank=False, check_finite=True,
     # backwards compatible only atol and rtol are both missing
-    if (rcond or cond) and (atol is None) and (rtol is None):
-        atol = rcond or cond
+    if ((rcond is not _NoValue or cond is not _NoValue)
+        and (atol is None) and (rtol is None)):
```
Indentation of the second line here is off.

h-vetinari commented on scipy/linalg/_basic.py:
```diff
@@ -1435,14 +1436,15 @@ def pinv(a, atol=None, rtol=None, return_rank=False, check_finite=True,
-        atol = rcond or cond
+    if ((rcond is not _NoValue or cond is not _NoValue)
+        and (atol is None) and (rtol is None)):
+        atol = rcond if rcond is not _NoValue else cond
```
I don't think this is correct for all combinations of `None`,  `_NoValue`, <else>.

j-bowhay replied:
Thanks good catch

h-vetinari commented on scipy/linalg/tests/test_basic.py:
```diff
@@ -1440,6 +1440,12 @@ def test_atol_rtol(self):
         assert_allclose(np.linalg.norm(adiff1), 4.233, rtol=0.01)
         assert_allclose(np.linalg.norm(adiff2), 4.233, rtol=0.01)
+    
+    def test_deprecation(self):
+        with pytest.deprecated_call(match='"cond" and "rcond"'):
+            pinv(np.ones((2,2)), cond=1)
+        with pytest.deprecated_call(match='"cond" and "rcond"'):
+            pinv(np.ones((2,2)), rcond=1)
```
Test should be parametrized and exercise all combinations (you can pass `_NoValue` explicitly to simulate the argument not being specified).

h-vetinari commented on scipy/linalg/_basic.py:
```diff
@@ -1435,14 +1436,15 @@ def pinv(a, atol=None, rtol=None, return_rank=False, check_finite=True,
 
-    if rcond or cond:
+    if rcond is not _NoValue or cond is not _NoValue:
         warn('Use of the "cond" and "rcond" keywords are deprecated and '
-             'will be removed in future versions of SciPy. Use "atol" and '
+             'will be removed in SciPy 1.14.0. Use "atol" and '
              '"rtol" keywords instead', DeprecationWarning, stacklevel=2)
```
Given the complexity of the whole tol/atol/rtol operation, and the fact that we were warning on this consistently already, I don't think we _need_ the full two releases here. That said, #18488 promises removal by 1.14 as well, so I'm fine with committing 1.14 in this PR. Basically I wanted to leave the door open for us to do it earlier if we feel the trade-off is worth it.

h-vetinari reviewed (commented):
Thanks for the ping. Some comments. I'm not a fan of pushing out the actual removal out by another 1.5 years (when 1.14 is released), but I guess that's what it is.

tupui commented:
> Thanks for the ping. Some comments. I'm not a fan of pushing out the actual removal out by another 1.5 years (when 1.14 is released), but I guess that's what it is.

I personally agree. We could maybe be more flexible with such things. At least allowing to deprecate in n+1 would be nice.

j-bowhay commented:
Thanks for the review @h-vetinari, I have hopefully addressed your comments. My personal preference would be to be conservative with the removal version. Six months does not make that much difference at our end but my experience is that often in industry it can take a long time to get security clearance for new versions of software and hence versions are sometimes missed.

ilayn commented:
I just realized that in #13831 I took the liberty to push `cond`, `rcond` to the end like an idiot. I don't know how such a change went under the bridge, but it even got some encouragement in my defense #18166 

 I apologize for my past behavior but as a silver lining nobody cared. Hence we might be overdoing this disciplined deprecation in a slightly overzealous fashion.  Obviously I'm not advocating such behavior but I think we need to consider the usage patterns on a case by case basis.

ilayn commented:
Ah now I remember, because `atol`, `rtol` is a drop in replacement for `cond`, `rcond` in the positional argument case. Not so idiot apparently but still unnecessary risk I'd say.

j-bowhay commented:
> Hence we might be overdoing this disciplined deprecation in a slightly overzealous fashion. Obviously I'm not advocating such behavior but I think we need to consider the usage patterns on a case by case basis.

To clarify, the primary purpose of this PR is to add the version number at which we can get rid of `cond`/`rcond`. The change to `_NoValue` here is probably not strictly necessary but I think it is important to follow the best practice as newcomers tend to copy (and extrapolate) existing code.

ilayn commented:
All good from my side. Just wanted to mention that for the very long ongoing deprecation warnings we might be a bit more lax. Otherwise I fully agree.

j-bowhay commented:
The wisdom of the crowd seems to prefer 1.13.0 given the existing long-standing warning so have updated the messages

h-vetinari commented on scipy/linalg/tests/test_basic.py:
```diff
@@ -1441,6 +1442,12 @@ def test_atol_rtol(self):
         assert_allclose(np.linalg.norm(adiff1), 4.233, rtol=0.01)
         assert_allclose(np.linalg.norm(adiff2), 4.233, rtol=0.01)
 
+    @pytest.mark.parametrize(["cond", "rcond"],
+                             [(1, 1), (1, _NoValue), (_NoValue, 1)])
```
Can you parametrize them separately please, and use `[1, None, _NoValue]` for both?

h-vetinari reviewed (commented):
Nothing major, but a few remaining nits

h-vetinari commented on scipy/linalg/tests/test_basic.py:
```diff
@@ -1441,6 +1442,15 @@ def test_atol_rtol(self):
+    @pytest.mark.parametrize("cond", [1, None, _NoValue])
+    @pytest.mark.parametrize("rcond", [1, None, _NoValue])
+    def test_deprecation(self, cond, rcond):
+        if cond is _NoValue and rcond is _NoValue:
+            pinv(np.ones((2,2)), cond=cond, rcond=rcond)
+        else:
+            with pytest.deprecated_call(match='"cond" and "rcond"'):
+                pinv(np.ones((2,2)), cond=cond, rcond=rcond)
```
```suggestion
        if cond is _NoValue and rcond is _NoValue:
            # the defaults if cond/rcond aren't set -> no warning
            pinv(np.ones((2,2)), cond=cond, rcond=rcond)
        else:
            # at least one of cond/rcond has a user-supplied value -> warn
            with pytest.deprecated_call(match='"cond" and "rcond"'):
                pinv(np.ones((2,2)), cond=cond, rcond=rcond)
```

h-vetinari reviewed (approved):
LGTM thanks!

Only a non-binding suggestion for adding a bit of context to the test, but approved in any case.

tupui reviewed (approved):
LGTM, and you all seem happy 😃 thanks
